src/Runtime/PyOMCompileExecutionSession.cpp

```diff
@@ -23,40 +23,60 @@ SUPPRESS_WARNINGS_POP
 
 namespace onnx_mlir {
 
+// =============================================================================
+// Constructor
+
 PyOMCompileExecutionSession::PyOMCompileExecutionSession(
-    std::string inputFileName, std::string sharedLibPath, std::string flags,
-    bool defaultEntryPoint)
-    : onnx_mlir::PyExecutionSessionBase(sharedLibPath, defaultEntryPoint) {
+    std::string inputFileName, std::string flags, bool defaultEntryPoint,
+    bool reuseCompiledModel)
+    : onnx_mlir::PyExecutionSessionBase() /* constructor without Init */ {
+  // First compile the onnx file.
   this->inputFileName = inputFileName;
-  if (this->inputFileName.empty()) {
-    errorMessage = "No OMCompileExecuteSession was created with the input file "
-                   "name specified.";
-  }
+  if (this->inputFileName.empty())
+    throw std::runtime_error(reportLibraryOpeningError(inputFileName));
+
   const char *outputName, *errorMsg;
-  int64_t rc;
-  rc = omCompileFromFile(
-      inputFileName.c_str(), flags.c_str(), &outputName, &errorMsg);
-  if (rc == 0) {
-    // Compilation success: save output file name.
-    this->sharedLibPath = std::string(outputName);
-    // Empty error.
-    errorMessage = std::string();
-  } else {
-    // Compilation failure: save error message.
-    errorMessage = std::string(errorMsg);
-    // Empty output file name.
-    this->sharedLibPath = std::string();
+  if (reuseCompiledModel) {
+    // see if there is a model to reuse.
+    outputName = omCompileOutputFileName(inputFileName.c_str(), flags.c_str());
+    FILE *file = fopen(outputName, "r");
+    if (file)
+      // File exists, we are ok.
+      fclose(file);
+    else
+      // File does not exist, cannot reuse compilation.
+      reuseCompiledModel = false;
   }
+  if (!reuseCompiledModel) {
+    int64_t rc;
+    rc = omCompileFromFile(
+        inputFileName.c_str(), flags.c_str(), &outputName, &errorMsg);
+    if (rc != 0) {
+      // Compilation failure: save error message.
+      errorMessage = std::string(errorMsg);
+      // Empty output file name.
+      this->outputFileName = std::string();
+      throw std::runtime_error(reportCompilerError(errorMessage));
+    }
+  }
+  // Compilation success: save output file name.
+  this->outputFileName = std::string(outputName);
```

Comment:
According to your comment for `omCompileOutputFileName`, do we need to free `outputName` here?